### DISCUSSION TITLE
4k performance improvable ?

### DISCUSSION THREAD

**Discussion by grobalt** [AmusementClub/vs-mlrt#19] [General]
https://www.svp-team.com/forum/viewtopic.php?pid=81483#p81483 integrated this as a test  -> in 1080p transcoding i reached around 115fps with RTX4090 and 22 fps for 2160p .. any idea to improve 4k performance ? :)

**Comment by WolframRhodium** [comment]
If tunings like increasing `num_threads`, setting `use_cuda_graph` are not useful, it is likely that the performance is limited by PCIe bandwidth rather than GPU compute. RIFE 4.6 (2x) alone should run at ~53.2 fps in 4k with 4090.

This problem is specific to rife in vs-mlrt and I will try to solve it. Thanks for the information!

Anyway, for advanced usage, it is possible to build a single trt model for multiple video resolutions by setting parameters `opt_shapes`, `max_shapes` and `static_shape` in `Backend.TRT`, but the performance is usually slower.

**Comment by grobalt** [comment]
Need to push this to SVP team as i dont know where to check the parameters num_threads etc. But i can confirm, i run the 4090 on a pcie4 x16 slot and another user confirms my framerate

**Comment by WolframRhodium** [comment]
We need additional information to determine why the performance degrades. It would be better if you could provide profiling information like [nsight system](https://developer.nvidia.com/nsight-systems) data. The installer can be downloaded without registration from the cuda toolkit.

**Reply by grobalt** [reply]
thanks, i think i found the file that set the parameter: vsmlrt.py
which includes:

`    @dataclass(frozen=False)
    class TRT:
        """ backend for nvidia gpus

        basic performance tuning:
        set fp16 = True (on RTX GPUs)
        increase num_streams
        increase workspace
        set use_cuda_graph = True
        """

        max_shapes: typing.Optional[typing.Tuple[int, int]] = None
        opt_shapes: typing.Optional[typing.Tuple[int, int]] = None
        fp16: bool = False
        device_id: int = 0
        workspace: int = 128
        verbose: bool = False
        use_cuda_graph: bool = False
        num_streams: int = 1
        use_cublas: bool = False # cuBLAS + cuBLASLt
        static_shape: bool = True
        tf32: bool = True
        log: bool = True

        # as of TensorRT 8.4, it can be turned off without performance penalty in most cases
        use_cudnn: bool = True
        use_edge_mask_convolutions: bool = True
        use_jit_convolutions: bool = True

        heuristic: bool = False # only supported on Ampere+ with TensorRT 8.5+

        # internal backend attributes
        supports_onnx_serialization: bool = False
`

but playing with streams 2 and cuda graph True did not change any performance .. maybe i am looking in the wrong place. currently downloading nsight system

**Comment by AkarinVS** [comment]
2160p video has 4x more data than 1080p. If your system can do 115fps at 1080p, then the max performance it can get for 2160p is 115/4 fps, which is around 28.75 fps.

You also need to consider the data movement cost, at 4k, each 22fps, the GPU requires 7.74 GB data from system DRAM and outputs 2.11GB. How much DRAM bandwidth do you have in your system? Even if it's not limited by PCIe bandwidth, it will very likely hit DRAM bandwidth ceiling. Remember that RIFE is not the only thing that uses the DRAM bandwidth, the rest of the VS script and SVP also consume DRAM bandwidth.

**Reply by grobalt** [reply]
Standard Ryzen system with 32 GB memory, dual rank 2 dimms.
The benchmarks https://github.com/AmusementClub/vs-mlrt/wiki/NVIDIA-GeForce-RTX-4090 show way more than i have, trying to figure out why (as my numbers are confirmed by other SVP users)

**Reply by AkarinVS** [reply]
I think it's just overhead from the rest of the VS script and SVP itself.

We have just benchmarked RIFE **v4.6** on 2160p input with RTX 4090 (`num_streams=2`), and got 53 fps (The wiki shows result for RIFE **v4.4**). But note our benchmark only test the speed of RIFE alone, no source filter or any downstream filters. As I've mentioned, even a single filter will require 2GB/s DRAM bandwidth at 22 fps.

**Reply by grobalt** [reply]
So there is still enough bandwidth available. I think i struggle in setting up the parameters correctly .. .

how can i do the benchmark you did ? sorry .. i don't have developer skills

**Comment by WolframRhodium** [comment]
Launch a new svp process by specifying path to svp executable in nsight system, and then transcode for a short period of time. Collection of GPU metrics should be enabled.

**Reply by grobalt** [reply]
done -> report uploaded https://drive.google.com/drive/folders/1EMsewBfnbVhFmmE9hgCvWGQAFk9FzHWj?usp=sharing

not transcoding, using and monitoring MPV directly

**Reply by grobalt** [reply]
from what is see ... target for 24p frame doubline is a "render time" below 20ms -> 48*20 = 960ms.

![image](https://user-images.githubusercontent.com/79084968/211192049-53b42e6f-3a87-4c72-9375-4ca544812c1c.png)

**Reply by WolframRhodium** [reply]
Thanks for the report. What's the value of `threads` of svp?

Ideally, memory copies should overlap computation, and the performance would be even higher. However it seems that the GPU is idle most of the time.

**Reply by WolframRhodium** [reply]
Is it possible to grab the vapoursynth script for isolated analysis? Probably mpv has some overheads.

**Reply by grobalt** [reply]
> Thanks for the report. What's the value of `threads` of svp?
> 
> Ideally, memory copies should overlap computation, and the performance would be even higher. However it seems that the GPU is idle most of the time.

the GPU thread setting in SVP is 1,2 or 4 -> all behave the same. Most likely no influence to the script.

**Reply by grobalt** [reply]
> Is it possible to grab the vapoursynth script for isolated analysis? Probably mpv has some overheads.

i think it uses the file C:\Program Files (x86)\SVP 4\script\base.py 
[base.zip](https://github.com/AmusementClub/vs-mlrt/files/10368280/base.zip)

**Reply by AkarinVS** [reply]
As clearly shown in the profile, the issue is not that RIFE is not fast enough, but mpv/SVP simply can't feed frame requests to the filter fast enough (see the next post for a profile where the GPU is fully saturated). In fact, the GPU sits idle for at least half the time.... In this particular case, no amount of tuning of vs-mlrt parameters can speed things up. One must find the real bottleneck elsewhere.

**Comment by WolframRhodium** [comment]
![Capture](https://user-images.githubusercontent.com/16359228/211195823-e0b63d23-0a65-427c-a31b-11456adda66f.PNG)
Here's the tuning target, in which computations overlap memory copies and the PCIe bandwidth is saturated. (`num_streams=5, use_cuda_graph=True` on A100, 108 fps at 4k)

[report2.zip](https://github.com/AmusementClub/vs-mlrt/files/10368386/report2.zip)

**Reply by grobalt** [reply]
Here new report with a new script
[32258583.zip](https://github.com/AmusementClub/vs-mlrt/files/10368466/32258583.zip)

https://drive.google.com/drive/folders/1EMsewBfnbVhFmmE9hgCvWGQAFk9FzHWj?usp=share_link
report4

**Reply by WolframRhodium** [reply]
The .py script is not complete and some constants are missing. After these params are filled, the script should be runnable by the command `vspipe.exe xxxx.py .` and this command should be profiled.

**Reply by WolframRhodium** [reply]
Anyway, the optimal value of `num_streams` depends on hardware configuration, video resolution and the model itself, but higher values should saturate the system more. Note that the value of `num_streams` should not be greater than `core.num_threads` in .py, otherwise resources are wasted.

**Comment by hooke007** [comment]
Based on my personal experience, mpv's vapoursynth filter is not fast enough. Even though the simpliest script would make 4K-video's playback laggy on my 10875H i7 CPU.
```
video_in.set_output()
```

**Comment by aloola18** [comment]
I'm using 5900x+3070ti and got 
30FPS with 3840x2176@30 > 60 with SVP transcoding.
~27fps with 3840x2176@25 > 50 with SVP transcoding.

 so it's clearly OP's setup has problems.
maybe PSU is not strong enough? which PSU are you using?

**Reply by grobalt** [reply]
i dont think you use rife with these numbers :) my numbers are the same as for other SVP users with 4090 in the SVP forum ...

**Comment by WolframRhodium** [comment]
Could you please run a baseline benchmark?

test.vpy (run with `vspipe -p test.vpy .`):
```python3
import vapoursynth as vs
from vsmlrt import RIFE, Backend, RIFEModel
core = vs.core


num_streams = 2
use_cuda_graph = True
peak_performance = True


if peak_performance:
    src = core.std.BlankClip(width=3840, height=2176, length=2000, format=vs.RGBS, color=[0.5]*3)
else:
    src = core.std.BlankClip(width=3840, height=2160, length=2000, format=vs.YUV420P8, color=[128]*3)
    src = core.misc.SCDetect(src)
    src = core.resize.Bicubic(src, format=vs.RGBS, matrix_in_s="709")
    src = core.std.AddBorders(src, 0, 0, 0, 16)

backend = Backend.TRT(fp16=True, num_streams=num_streams, use_cuda_graph=use_cuda_graph)
flt = RIFE(src, multi=2, model=RIFEModel.v4_6, backend=backend)

if not peak_performance:
    flt = core.std.Crop(flt, 0, 0, 0, 16)

flt.set_output()
```

`peak_performance=True` is an ideal case.

It seems that `use_cuda_graph=True` is more useful for TCC driver mode (for professional cards) than WDDM.

**Reply by hooke007** [reply]
Here is my test results (could be as additional ref.)

First to find the reasonable value for my RTX 4080 with `use_cuda_graph = True`  `peak_performance = True`
`num_streams=5` Output 4000 frames in 115.37 seconds (34.67 fps)
`num_streams=3` Output 4000 frames in 113.38 seconds (35.28 fps)
`num_streams=2` Output 4000 frames in 114.47 seconds (34.94 fps)

Then use 2 streams to test `peak_performance = False`
Output 4000 frames in 141.55 seconds (28.26 fps)

edit: After overclocking my memory from 4800MHz to 5600Mhz, I got the new result: Output 4000 frames in 127.88 seconds (31.28 fps)

**Reply by WolframRhodium** [reply]
Thanks!

**Reply by grobalt** [reply]
`C:\Program Files (x86)\SVP 4\script>vspipe -p test.vpy .
Script evaluation failed:
Python exception: No attribute with the name ncnn exists. Did you mistype a plugin namespace?

Traceback (most recent call last):
  File "C:\Program Files (x86)\SVP 4\rife\vsmlrt.py", line 34, in get_plugins_path
    path = core.trt.Version()["path"]
  File "src\cython\vapoursynth.pyx", line 2434, in vapoursynth._CoreProxy.__getattr__
  File "src\cython\vapoursynth.pyx", line 2308, in vapoursynth.Core.__getattr__
AttributeError: No attribute with the name trt exists. Did you mistype a plugin namespace?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files (x86)\SVP 4\rife\vsmlrt.py", line 37, in get_plugins_path
    path = core.ort.Version()["path"]
  File "src\cython\vapoursynth.pyx", line 2434, in vapoursynth._CoreProxy.__getattr__
  File "src\cython\vapoursynth.pyx", line 2308, in vapoursynth.Core.__getattr__
AttributeError: No attribute with the name ort exists. Did you mistype a plugin namespace?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files (x86)\SVP 4\rife\vsmlrt.py", line 40, in get_plugins_path
    path = core.ov.Version()["path"]
  File "src\cython\vapoursynth.pyx", line 2434, in vapoursynth._CoreProxy.__getattr__
  File "src\cython\vapoursynth.pyx", line 2308, in vapoursynth.Core.__getattr__
AttributeError: No attribute with the name ov exists. Did you mistype a plugin namespace?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "src\cython\vapoursynth.pyx", line 2866, in vapoursynth._vpy_evaluate
  File "src\cython\vapoursynth.pyx", line 2867, in vapoursynth._vpy_evaluate
  File "test.vpy", line 2, in <module>
    from vsmlrt import RIFE, Backend, RIFEModel
  File "C:\Program Files (x86)\SVP 4\rife\vsmlrt.py", line 48, in <module>
    plugins_path: str = get_plugins_path()
  File "C:\Program Files (x86)\SVP 4\rife\vsmlrt.py", line 42, in get_plugins_path
    path = core.ncnn.Version()["path"]
  File "src\cython\vapoursynth.pyx", line 2434, in vapoursynth._CoreProxy.__getattr__
  File "src\cython\vapoursynth.pyx", line 2308, in vapoursynth.Core.__getattr__
AttributeError: No attribute with the name ncnn exists. Did you mistype a plugin namespace?


C:\Program Files (x86)\SVP 4\script>`

**Reply by AkarinVS** [reply]
please add the following lines to the front of the script and try again:
```python
import vapoursynth as vs
core = vs.core
if not hasattr(core,'trt'):
    core.std.LoadPlugin("C:\\Program Files (x86)\\SVP 4\\rife\\vstrt.dll")
import sys 
sys.path.append("C:\\Program Files (x86)\\SVP 4\\rife")
```

**Reply by grobalt** [reply]
Thanks a lot AkarinVS, that was quick.

Results:
C:\Program Files (x86)\SVP 4\script>vspipe -p test.vpy .
vstrt: TensorRT version mismatch, built with 2135 but loaded with 2136; continue but fingers crossed...
CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
CUDA lazy loading is not enabled. Enabling it can significantly reduce device memory usage. See `CUDA_MODULE_LOADING` in https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars
Script evaluation done in 0.70 seconds
Output 4000 frames in 67.73 seconds (**59.05 fps**)

Let me post the building process as well (as i deleted the model before running the first time ... and maybe you see something odd here):
[4000framebenchmark.txt](https://github.com/AmusementClub/vs-mlrt/files/10371250/4000framebenchmark.txt)

Nsight file with num_streams = 2
https://drive.google.com/file/d/1Uwr1Kz8lfsh9wof6NLQMJMya_3y8s6sb/view?usp=share_link

num_streams = 1 results in **34,81fps**
num_streams = 2 results in ~**59fps**
num_streams = 3 results in **59,32fps**
num_streams = 4 results in ~**59fps**

playing with the next parameter:
num_streams = 2
use_cuda_graph = False
peak_performance = False
results in **37,7fps**

num_streams = 2
use_cuda_graph = True
peak_performance = False
results in **37,7fps**

num_streams = 2
use_cuda_graph = False
peak_performance = True
results in **58fps**

**Reply by chainikdn** [reply]
what is the CPU? memory in a dual channel mode?
add
`core.num_threads = 8` after `core = vs.core` and play with this number: 8, 12, 16, 24
... with peak_performance = False, obviously

**Reply by grobalt** [reply]
Ryzen 5600x, CPU has about 17% load
Why peak_performance false ? True was way faster .. confused

**Reply by chainikdn** [reply]
Because "False" is how it will be a real life and "True" is just a synthetic test.
It's not about CPU load, it's about memory throughput. Increasing CPU threads may help.
Memory in a dual channel mode?

**Reply by grobalt** [reply]
> Because "False" is how it will be a real life and "True" is just a synthetic test. It's not about CPU load, it's about memory throughput. Increasing CPU threads may help. Memory in a dual channel mode?


here results with peak_performance = True
core.num_threads = 6 (as my 5600x is a 6-core):  59fps
core.num_threads = 8:  54fps
core.num_threads = 12 (6+6HT):  58,94fps

here results with peak_performance = False
core.num_threads = 6 (as my 5600x is a 6-core):  37,75fps
core.num_threads = 12 (6+6HT):  38,06fps

**Reply by MysteryDove** [reply]
I'me getting 34.6fps for peak_performance = false and 56.1fps for peak_performance = true

From what I can tell monitoring DRAM Read/Write using HWInfo while running the script, I have to say the fps is totally bottlenecked by the system dram.
I'm getting both around 45Gbps for Read+Write when running the script. When I using memory crunching software like TM5, I'll get  around 45Gbps for DRAM R/W.

**Reply by grobalt** [reply]
<img width="413" alt="image" src="https://user-images.githubusercontent.com/79084968/211268958-5de963da-9dc2-4ef5-a297-3841a3cf921a.png">
results are valid for the single CCD Ryzen 5600x

**Reply by grobalt** [reply]
> I'me getting 34.6fps for peak_performance = false and 56.1fps for peak_performance = true
> 
> From what I can tell monitoring DRAM Read/Write using HWInfo while running the script, I have to say the fps is totally bottlenecked by the system dram. I'm getting both around 45Gbps for Read+Write when running the script. When I using memory crunching software like TM5, I'll get around 45Gbps for DRAM R/W.

hwinfo shows "halfed" numbers for DDR, correct ?
<img width="466" alt="image" src="https://user-images.githubusercontent.com/79084968/211279427-58db0da2-6e22-4f40-b40d-f8c7dd2d8466.png">

Then i am at the limit as well ..  
i updated bios and gained a bit better latency but not a big step.

**Reply by WolframRhodium** [reply]
I got 75fps on epyc zen2 + 3090. (AIDA64 BW ~120 GB/s, `peak_performance=True, num_streams=3, use_cuda_graph=1`)

**Reply by grobalt** [reply]
which cpu of zen2 is doable 120GB/s ? threadripper ?

**Reply by WolframRhodium** [reply]
server class (epyc). I don't know the exact model in a virtualized environment.

**Reply by grobalt** [reply]
ok .. if that is needed i would grab my server that is in a friens datacenter , that one has 190GB/s (7713 epyc with 8 dimms)
Anyhow, just ordered DDR5 7200, Ryzen 7700x and a new mainboard  -> should be in the range of 100 GB/s this wednesday ;)

**Reply by chainikdn** [reply]
PyTorch/vsrife eats RGBH instead of RGBS. This would save some ram bandwidth, I guess...

**Reply by MysteryDove** [reply]
To support my hypothesis about memory bottlenecking, I've tested with ddr4-2133 and I got significantly lower numbers. Regardless of memory timing, I've got 56.26 fps under ddr4-3600 and 42.43fps with ddr4-2133 which is a 24.6% perf decrease.

Also, while have some profiling with [WolframRhodium](https://github.com/WolframRhodium), we've found the Hardware-Accelerated GPU Scheduling in Windows 11 is actually affecting performance, here's some test results:

none: 56.26
with rbar: 56.16
with hardware scheduler: 54.38
with hardware scheduler+rbar: 55.39

**Reply by chainikdn** [reply]
So, all _you_ can do about it is allowing vs.RGBH input when fp16=True. 
Script-level / vapoursynth / mpv optimizations are out of scope here. For example, putting AddBorders before Resize should help a bit...

**Reply by AkarinVS** [reply]
before embarking on allowing fp16, we should first identify and understand the bottleneck the prevented mpv/svp from approaching ~37fps (currently only ~22fps).

From the GPU profiling graph, it's clear the GPU is idle more than half of the time, so even though the vapoursynth version is memory bound, I'm not convinced that the OP's mpv performance is memory bound as well.

**Reply by chainikdn** [reply]
Your test script is missing reverse conversion from RGBS to YUV420. Plus there're two more frames copying inside mpv - from mpv to vapoursynth and then back. Plus a copy-back video decoder, obviously.
Anyway, one RGBS frame is 8 times bigger than YUV420P8 one.

**Reply by AkarinVS** [reply]
Here is the script that adds reverse RGBS to YUV conversion & compensate for extra frame copying:

(Non-SVP users remove the `core.std.LoadPlugin` line.)
```python
import vapoursynth as vs
from vsmlrt import RIFE, Backend, RIFEModel
core = vs.core

if not hasattr(core,'trt'):
    core.std.LoadPlugin("C:\\Program Files (x86)\\SVP 4\\rife\\vstrt.dll")
import sys 
sys.path.append("C:\\Program Files (x86)\\SVP 4\\rife")

num_streams = 2
use_cuda_graph = True
peak_performance = False

if peak_performance:
    src = core.std.BlankClip(width=3840, height=2176, length=2000, format=vs.RGBS, color=[0.5]*3)
else:
    src = core.std.BlankClip(width=3840, height=2160, length=2000, format=vs.YUV420P8, color=[128]*3)
    src = core.misc.SCDetect(src)
    src = core.resize.Bicubic(src, 3840, 2176, format=vs.RGBS, matrix_in_s="709", src_height=2176)

backend = Backend.TRT(fp16=True, num_streams=num_streams, use_cuda_graph=use_cuda_graph)
flt = RIFE(src, multi=2, model=RIFEModel.v4_6, backend=backend)

if not peak_performance:
    flt = flt.resize.Bicubic(3840, 2160, format=vs.YUV420P8, matrix_s="709", src_height=2160)

    # force two more frame copies to compensate for mpv <-> vs overhead.
    flt = core.std.Expr(flt, "x 1 +") # it's actually slower than frame copying
    flt = core.std.Expr(flt, "x 1 -")

flt.set_output()
```

**Reply by hooke007** [reply]
If you were talking about playback,
WM4 said mpv's vs filter is a gross hack, more info ref https://github.com/mpv-player/mpv/issues/7020#issuecomment-537919816.

Even though running the similar script, mpv's rendering speed is much slower than vseditor.
Such a simple script https://github.com/AmusementClub/vs-mlrt/discussions/19#discussioncomment-4625941 would make the video's playback terrrible in my old device.

**Reply by chainikdn** [reply]
> WM4 said mpv's vs filter is a gross hack

And I personally know how to make it even more gross and hacky :D and eliminate excessive frames copying back and forth, but I won't even make a PR for that, cause they will never accept it anyway...
But again, this is out of scope here.

**Reply by WolframRhodium** [reply]
[v12.3.test](https://github.com/AmusementClub/vs-mlrt/releases/tag/v12.3.test) release adds support for fp16 I/O.

You only need to update files in [
scripts.v12.3.test.7z](https://github.com/AmusementClub/vs-mlrt/releases/download/v12.3.test/scripts.v12.3.test.7z) and [VSTRT-Windows-x64.v12.3.test.7z](https://github.com/AmusementClub/vs-mlrt/releases/download/v12.3.test/VSTRT-Windows-x64.v12.3.test.7z), and the benchmarking script has been updated with parameters `input_format` and `output_format`.

```python3
import vapoursynth as vs
from vsmlrt import RIFE, Backend, RIFEModel
core = vs.core

if not hasattr(core,'trt'):
    core.std.LoadPlugin("C:\\Program Files (x86)\\SVP 4\\rife\\vstrt.dll")
    import sys 
    sys.path.append("C:\\Program Files (x86)\\SVP 4\\rife")

num_streams = 2
use_cuda_graph = True
peak_performance = False
input_format = 0 # 0: fp32, 1: fp16
output_format = 0 # 0: fp32, 1: fp16

if peak_performance:
    src = core.std.BlankClip(width=3840, height=2176, length=2000, format=vs.RGBS if input_format == 0 else vs.RGBH, color=[0.5]*3)
else:
    src = core.std.BlankClip(width=3840, height=2160, length=2000, format=vs.YUV420P8, color=[128]*3)
    src = core.misc.SCDetect(src)
    src = core.resize.Bicubic(src, 3840, 2176, format=vs.RGBS if input_format == 0 else vs.RGBH, matrix_in_s="709", src_height=2176)

backend = Backend.TRT(fp16=True, num_streams=num_streams, use_cuda_graph=use_cuda_graph, output_format=output_format)
flt = RIFE(src, multi=2, model=RIFEModel.v4_6, backend=backend)

if not peak_performance:
    flt = flt.resize.Bicubic(3840, 2160, format=vs.YUV420P8, matrix_s="709", src_height=2160)

    # force two more frame copies to compensate for mpv <-> vs overhead.
    flt = core.std.Expr(flt, "x 1 +") # it's actually slower than frame copying
    flt = core.std.Expr(flt, "x 1 -")

flt.set_output()
```

**Reply by chainikdn** [reply]
rife_v4.6.onnx.1920x1088_fp16_workspace256_trt-8502_cudnn_I-fp16_O-fp32_NVIDIA-GeForce-RTX-2060_3dcbe72f.engine.cache.lock
"I-fp16_O-fp32" part looks suspicious

File 'C:\Program Files (x86)\SVP 4\rife\vsmlrt.py', line 948, in RIFE
return core.std.Interleave([clip, output])
Interleave: clips are mismatched in format starting at clip #1, passed Undefined[1920x1088] and RGBS[1920x1088]

**Reply by WolframRhodium** [reply]
Thanks, please update files again.

**Reply by hooke007** [reply]
Have no idea what happened but I got the impressing result. ref https://github.com/AmusementClub/vs-mlrt/discussions/19#discussioncomment-4629367

in&out fp32
Output 4000 frames in 72.82 seconds (54.93 fps)

in&out fp16
Output 4000 frames in 52.98 seconds (75.50 fps)

edit: I checked it out. input fp32 with output fp16 would seriously make speed much slow (Output 4000 frames in 107.41 seconds (37.24 fps))

**Reply by chainikdn** [reply]
can't say anything other than "it works" cause I'm GPU-capped and there's no perf. difference in this case

**Reply by WolframRhodium** [reply]
don't know why input fp32 with output fp16 is slow

**Comment by hooke007** [comment]
Graet work!
Now RTX4080 could make 24fp→48fps(3840X2160yuv444p12→p16) smooth in mpv.
<img src="https://user-images.githubusercontent.com/41094733/211379880-1ee99bc3-05f5-4002-b794-9ad01a36f017.jpg" height=500>

**Reply by grobalt** [reply]
this is non HDR, correct ? :) with HDR it is still not fast enough

**Reply by hooke007** [reply]
common hdr is only 10bit.
Tonemapping is another gpu-intensive thing, but it is in the end of processing stage in vo_gpu, while `vf` has to work before any user hooks. It's out of vapoursynth's control.

**Reply by grobalt** [reply]
for me hdr is working as well .. but somehow ... an LG test clip is not working and i dont understand why not

https://4kmedia.org/lg-new-york-hdr-uhd-4k-demo/

<img width="1229" alt="hdr10" src="https://user-images.githubusercontent.com/79084968/211413250-66b18c72-e909-4f2b-b720-7237127862e1.png">

**Reply by hooke007** [reply]
Smooth here. Could be another issue on your side.
<img src="https://user-images.githubusercontent.com/41094733/211414614-b0b33ff3-1b6c-45ba-bdcc-4394ec8e717a.jpg" height=200>

**Reply by grobalt** [reply]
2 rife threads ? 10875H i7 CPU ?

**Reply by hooke007** [reply]
Nope, that's my obsolete device (10875H & 2070s). I use 7950x & 4080 now.

**Reply by grobalt** [reply]
Ok .. 7950 with good memory has nearly double memory bandwidth than my Ryzen 5600x with 3800Mhz mem and 2 CCD as well

**Comment by aloola18** [comment]
I just want to report that turning OFF Hardware-accelerated GPU greatly increases FPS on Windows 11.

ON
https://cdn.discordapp.com/attachments/290709370600423424/1066759015536799855/image.png

OFF
https://cdn.discordapp.com/attachments/290709370600423424/1066757184819576882/image.png

**Reply by hooke007** [reply]
use x264/265 instead, unless you are sure that "Hardware-accelerated" doesn't affect nvenc.

**Reply by aloola18** [reply]
> use x264/265 instead, unless you are sure that "Hardware-accelerated" doesn't affect nvenc.

sorry I mean Hardware-accelerated GPU scheduling windows 11. 
turn it off in  System > display > graphics > default graphics settings >

**Reply by hooke007** [reply]
I know what you were talking about.
I mean you'd better to use vspipe only to test. Because you don't know if "Hardware-accelerated" would affect other processing stages.

**Reply by aloola18** [reply]
The old script I saw above:
OFF: Output 4000 frames in 104.49 seconds (38.28 fps)
ON: Output 4000 frames in 108.95 seconds (36.71 fps)

edited the script output with FP16-FP16
OFF: Output 4000 frames in 75.87 seconds (52.72 fps)
ON: Output 4000 frames in 85.03 seconds (47.04 fps)

**Reply by WolframRhodium** [reply]
We observe similar improvements here. It seems nv's hardware scheduling implementation reduces the number of concurrent tasks for reduced latency, but throughout is sacrificed.

**Comment by dtlnor** [comment]
here is some result I tested with vsmlrt v13.1

I use the script above and use vspipe to benchmark. It gives very weird result. Idk if this is problem from my end.

## vsmlrt v13.1
- AMD Ryzen™ 9 7950X
- DDR5 2x32 6000mhz
- Nvidia 4090 (wddm, ecc disabled, rBar enabled)
- GPU processor clock @ 2490 MHz
- GPU memory clock @ 1469 MHz
- driver 528.49
- CUDA driver: 12.0.147
- Windows 10 22H2
- Hardware accelerated GPU scheduling OFF
- VapourSynth R60 (py 3.10 64bit)
- backend:
    `fp16 = True,
    use_cuda_graph = True` ( using the script above)


| model              |   | TRT 1 stream  | TRT 2 streams |
| ---                | - | ---           | ---           |
| rife io-fp16 peak static <br>(3840x2176, model=46) |   | 64.06 / 1706<br>(report 14, ~68% usage) | 123.22 / 2720<br>(report 13, ~100% usage) |
| rife io-fp16 overhead static <br>(3840x2176, model=46) |   | 56.41 / 1680<br>(report 15, ~60% usage) | 68.37 / 2721<br>(report 16, ~80% usage) |
| rife io-fp32 peak static <br>(3840x2176, model=46) |   | 31.99 / 1974<br>(report 19, ~50% usage) | 49.73 / 3327<br>(report 18 , ~70% usage) |
| rife io-fp32 overhead static <br>(3840x2176, model=46) |   | 37.01 / 1983<br>(report 22, ~60% usage) | 53.92 / 3381<br>(report 21 , ~80% usage) |

some reports if anyone interested: https://drive.google.com/drive/folders/1zhigwILaQm13FgY4O5w7oZ18iABY5Lhw?usp=share_link

only rife io-fp16 peak performance has reach 100% cuda usage and gives > 120fps result. Others just not able to.  
Also there is a weird case is for io-fp32 peak, I had reach 74.59fps once, with 100% cuda usage(I didn't do profiling with that run).  
I have never reach 70fps again since that one.

note: I didn't flip the result of io-fp32 peak and overhead, the peak performance one is slower than the mpv simulate one. I have no idea why.

**Reply by WolframRhodium** [reply]
RIFE performance on multi-chiplet processors may be affected by inter-chiplet bandwidth. Please consider to lower the value of `core.num_threads` to 8, for example, or other thread pinning techniques.

**Reply by MysteryDove** [reply]
We've monitored huge performance hit when running on a 5950x cpu with no thread limit, please consider bind the workload to single ccd by lowering core.num_threads or other software

**Reply by dtlnor** [reply]
@MysteryDove @WolframRhodium Ah, I miss that statement. Thank you for your helps. The value between 4~16 are stable for me (~73fps), not setting the core.num_threads or over 28 are likely to taking ~49fps

**Reply by dtlnor** [reply]
Here is my new result with `core.num_threads` = 16. Its OKey new. You guys are awesome.
| model              |   | TRT 1 stream  | TRT 2 streams |
| ---                | - | ---           | ---           |
| rife io-fp16 peak static <br>(3840x2176, model=46) |   | 63.34 / 1687 | 121.22 / 2708 |
| rife io-fp16 overhead static <br>(3840x2176, model=46) |   | 55.65 / 1695 | 69.14 / 2706 |
| rife io-fp32 peak static <br>(3840x2176, model=46) |   | 39.06 / 2110 | 74.39 / 3331 |
| rife io-fp32 overhead static <br>(3840x2176, model=46) |   | 36.73 / 1999 | 52.09 / 3312 |
